From 4f1298263ef0ad93843e306f4f1b7bb7db8cc40b Mon Sep 17 00:00:00 2001
From: Tomas Frydrych <tomas@sleepfive.com>
Date: Wed, 5 Sep 2012 13:25:57 +0100
Subject: [PATCH] cogl-pipeline: fix memory corruption case freeing layer
 cache

Check that the layer cache is of non-zero size before calling g_slice_free.
---
 cogl/cogl-pipeline.c |    3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/cogl/cogl-pipeline.c b/cogl/cogl-pipeline.c
index 9f21c2f..38e7c22 100644
--- a/cogl/cogl-pipeline.c
+++ b/cogl/cogl-pipeline.c
@@ -254,7 +254,8 @@ recursively_free_layer_caches (CoglPipeline *pipeline)
   if (pipeline->layers_cache_dirty)
     return;
 
-  if (G_UNLIKELY (pipeline->layers_cache != pipeline->short_layers_cache))
+  if (G_UNLIKELY (pipeline->layers_cache != pipeline->short_layers_cache &&
+                  pipeline->n_layers != 0))
     g_slice_free1 (sizeof (CoglPipelineLayer *) * pipeline->n_layers,
                    pipeline->layers_cache);
   pipeline->layers_cache_dirty = TRUE;
-- 
1.7.10.4

